src/Native/Runtime/arm/WriteBarriers.S

```diff
@@ -21,48 +211,135 @@ LEAF_END RhpAssignRef, _TEXT
 // WARNING: Code in EHHelpers.cpp makes assumptions about write barrier code, in particular:
 // - Function "InWriteBarrierHelper" assumes an AV due to passed in null pointer will happen on the first instruction
 // - Function "UnwindWriteBarrierToCaller" assumes the stack contains just the pushed return address
-LEAF_ENTRY RhpCheckedAssignRef, _TEXT
+LEAF_ENTRY RhpCheckedAssignRef\EXPORT_REG_NAME, _TEXT
 
-// ALTERNATE_ENTRY RhpCheckedAssignRef
+// Export the canonical write barrier under unqualified name as well
+.ifc \REFREG, r1
+ALTERNATE_ENTRY RhpCheckedAssignRef
 ALTERNATE_ENTRY RhpCheckedAssignRefAVLocation
+.endif
+
+          // Write the reference into the location. Note that we rely on the fact that no GC can occur between here
+          // and the card table update we may perform below.
+          str          \REFREG, [r0]
+
+          DEFINE_CHECKED_WRITE_BARRIER_CORE RhpCheckedAssignRef, \REFREG, \TRASHREG
+
+          bx           lr
+LEAF_END RhpCheckedAssignRef\EXPORT_REG_NAME, _TEXT
+.endm
+
+// One day we might have write barriers for all the possible argument registers but for now we have
+// just one write barrier that assumes the input register is RSI.
+DEFINE_CHECKED_WRITE_BARRIER r1, r1, r3
+
+// WARNING: Code in EHHelpers.cpp makes assumptions about write barrier code, in particular:
+// - Function "InWriteBarrierHelper" assumes an AV due to passed in null pointer will happen at RhpCheckedLockCmpXchgAVLocation
+// - Function "UnwindWriteBarrierToCaller" assumes the stack contains just the pushed return address
+// r0 = destination address
+// r1 = value
+// r2 = comparand
+LEAF_ENTRY RhpCheckedLockCmpXchg, _TEXT
+          PROLOG_PUSH "{r4,lr}"
```

Comment:
This does not need scratch registers - it should be like the Windows version: https://github.com/dotnet/corert/blob/master/src/Native/Runtime/arm/WriteBarriers.asm#L343